typing/typecore.ml

```diff
@@ -6786,18 +6785,28 @@ and type_expect_
         raise (Error (exp.exp_loc, env, Unsupported_stack_allocation category))
       in
       begin match exp.exp_desc with
-      | Texp_function { alloc_mode; _} | Texp_tuple (_, alloc_mode)
+      | Texp_function { alloc_mode; _} ->
+          submode ~loc ~env
+            (Value.min_with (Comonadic Areality) Regionality.local)
+            expected_mode;
+          begin match Locality.submode Locality.local
+            (Alloc.proj (Comonadic Areality) alloc_mode.mode) with
+          | Ok () -> ()
+          | Error e ->
+              raise (Error (exp.exp_loc, env,
+                Uncurried_function_escapes (Error (Comonadic Areality, e))))
+          end
+      | Texp_tuple (_, alloc_mode)
       | Texp_construct (_, _, _, Some alloc_mode)
       | Texp_variant (_, Some (_, alloc_mode))
       | Texp_record {alloc_mode = Some alloc_mode; _}
       | Texp_array (_, _, _, alloc_mode)
       | Texp_field (_, _, _, Boxing (alloc_mode, _), _) ->
-        begin match Locality.submode Locality.local
-          (Alloc.proj (Comonadic Areality) alloc_mode.mode) with
-        | Ok () -> ()
-        | Error _ -> raise (Error (e.pexp_loc, env,
-            Cannot_stack_allocate alloc_mode.locality_context))
-        end
+          submode ~loc ~env
+            (Value.min_with (Comonadic Areality) Regionality.local)
+            expected_mode;
+          Locality.submode_exn Locality.local
+            (Alloc.proj (Comonadic Areality) alloc_mode.mode)
```

Comment:
Are there any tests of this path? It looks like you've changed the error message, but I don't see a corressponding test output change.

Reply:
The original code type check `Texp_tuple` with expected_mode being local. That should make `Texp_tuple.alloc_mode` extracted here to be `local` and therefore the original `Cannot_stack_allocate` will never trigger. It's the other submode below that might fail.

The new code change the always-successful `submode` to `submode_exn`. The behavior doesn't change.

Reply:
Ugh, nvm, my assumption is wrong - I will fix it.

Reply:
Yeah my previously assumption was wrong. I've rewritten the PR and it's now much smaller. Please see the updated description and review by commit.